lib/resourcemerge/core.go

```diff
@@ -39,13 +52,25 @@ func ensurePodSpec(modified *bool, existing *corev1.PodSpec, required corev1.Pod
 		}
 		ensureContainer(modified, existingCurr, required)
 	}
-
+	// required is new
+	// existing is current
+	// so we want to add required NOT existing
+	// but since the names do not match, kube-rbac-proxy will be treated like a new ctr.
+	// so "existingCurr" will enter ensureContainer empty
+	// so, oauth-proxy and kube-rbac-proxy will never enter ensureContainer together
+	// instead we will add both since kube-rbac-proxy will be parsed seprately from oauth-proxy....
+	// and since we use existing as the source of truth to add to the new daemonset
+	// we add both oauth-proxy and kube-rbac-proxy
+	// so we need a way to detect if containers with different names are going to collide.
 	for _, required := range required.Containers {
 		var existingCurr *corev1.Container
 		for j, curr := range existing.Containers {
 			if curr.Name == required.Name {
 				existingCurr = &existing.Containers[j]
 				break
+			} else if ensureMinimalOverlap(&curr, &required) {
+				existingCurr = &existing.Containers[j]
+				break
```

Comment:
This is what is was generating before, btw, I had it dump it out: https://gist.github.com/jkyros/b23fecbff849fb391fddffa1fd36d3cf so yep, they were both in there and causing a conflict. 

If I understand this right, you're telling it that the `kube-rbac-proxy` container and the `oauth-proxy` pod are equivalent -- they "match" because they have a port conflict instead of because of their names -- which works in this specific case. 

In the general case though...where they didn't conflict, e.g. we just have a new manifest that adds a new container and removes an old container, we'd get the new container, but we'd still end up stuck with the old container, wouldn't we? (nothing ever removes old containers from the pod spec, does it?)

Reply:
the `ensureMinimalOverLap` block only executes if true (if the ports overlap, but we can add more cases). Let's say we have two containers, with completely different configs, one is old one is new. We would never enter the `else if` since the configs are different

I think you may be right that we never remove old containers... that seems a little problematic come to think of it. It seems that we merge where we can and now avoid overlap but if you have `randomCTR123` and it just so happens to not have any overlap with any new containers you are adding, it will just slip under the radar...

Reply:
>the ensureMinimalOverLap block only executes if true (if the ports overlap, but we can add more cases). Let's say we have two containers, with completely different configs, one is old one is new. We would never enter the else if since the configs are different

I just thought that was a weird corner that if two containers that were otherwise completely different specified the same port resourcemerge would merge them. The recipient of that future surprise would be...well...surprised :smile: 

> I think you may be right that we never remove old containers... that seems a little problematic come to think of it. It seems that we merge where we can and now avoid overlap but if you have randomCTR123 and it just so happens to not have any overlap with any new containers you are adding, it will just slip under the radar...

Yeah, I was thinking "If we cleaned up the existing ones properly or just overwrote the whole container section, the collision wouldn't have ever happened".  

Anyway, all we _really_ need to do is just not leave a minefield in the short term until we switch to the actual `library-go` (instead of our fork) and use[ its functions ](https://github.com/openshift/library-go/blob/add9892da47ef737135dae6a58e33127fd79ba70/pkg/operator/resource/resourceapply/apps.go#L181)-- which only really worry about the generation and merging the metadata and just overwrite everything else. 

(And if now you are like "then couldn't we just do something like upstream `library-go` does for now and overwrite existing.Containers with required.Containers if the lists are different", I think the answer is yes)